cw/cw305/correlation-enhanced_collision_attack.py

```diff
@@ -0,0 +1,407 @@
+
+        # Plot m_alpha_j.
+        if plot_debug:
+            plot_m_alpha_j = [[] for i in range(16 * 256)]
+            for j in range(16):
+                for alpha in range(256):
+                    plot_m_alpha_j[j * 256 + alpha] = m_alpha_j[j, alpha]
+
+            plot.save_plot_to_file(plot_m_alpha_j[0:3], 3, 'm_alpha_j.html')
+
+        ####################################################################
+        # Find maximum correlation of m_alpha_j for every pair of plain or #
+        # cipher-text positions a and b.                                   #
+        ####################################################################
+        max_rho_list = []
+        rho_list = []
+        for a in range(16):
+            for b in range(a + 1, 16):
+                max_rho, rho = get_max_rho(m_alpha_j, a, b)
+                max_rho_list.append(max_rho)
+                rho_list.append(rho)
+
+        # Plot correlation coefficients for delta of plain/cipher-text byte
+        # pairs.
+        if plot_rho:
+            plot.save_plot_to_file(rho_list[0:3], 3, 'rho.html')
+
+        # Convert into matrix.
+        max_rho_mat = [[[0, 0] for b in range(16)] for a in range(16)]
+        i = 0
+        for a in range(16):
+            for b in range(a + 1, 16):
+                max_rho_mat[a][b] = max_rho_list[i]
+                i += 1
+
+        # Create ordered list of relationships between byte positions.
+        max_rho_deltas = [[0, 0, 0, 0]]
+        for a in range(16):
+            for b in range(a + 1, 16):
+                rho = max_rho_mat[a][b][0]
+                delta = max_rho_mat[a][b][1]
+                for i in range(len(max_rho_deltas)):
+                    if rho > max_rho_deltas[i][0]:
+                        max_rho_deltas.insert(i, [rho, delta, a, b])
+                        break
+
+        # Remove the zero element.
+        del max_rho_deltas[-1]
+
+        #########################################
+        # Find the key through trial and error. #
+        #########################################
+
+        # We take have 120 deltas and take the 15 most promising ones to try
+        # out the remaining 256 key possibilties.
+        key_guess_bytes = np.zeros((16), np.uint8)
+        num_bytes_match = 0
+        plaintext = project.textins[0]
+        ciphertext = project.textouts[0]
+        for i in range(256):
+            key_temp = np.zeros((16), np.uint8)
+            num_bytes_match_temp = 0
+            status = np.zeros((16), np.bool_)
+            # Start with the most promising delta.
+            rho, delta, a, b = max_rho_deltas[0]
+            key_temp[a] = i
+            status[a] = True
+            key_temp[b] = i ^ delta
+            status[b] = True
+            for byte in range(15):
+                # Take the most promising delta that involves the available
+                # key bytes.
+                for rho, delta, a, b in max_rho_deltas:
```

Comment:
Would you mind adding a comment here saying that this is a single-pass heuristic? Since we choose the `delta` value for each byte sequentially from 0 to 15, we don't consider cases where byte x could use the delta value of byte y where y > x.